planner/cascades/optimize.go

```diff
@@ -129,16 +126,23 @@ func (opt *Optimizer) onPhasePreprocessing(sctx sessionctx.Context, plan planner
 }
 
 func (opt *Optimizer) onPhaseExploration(sctx sessionctx.Context, g *memo.Group) error {
-	for !g.Explored {
-		err := opt.exploreGroup(g)
-		if err != nil {
-			return err
+	for i, ruleBatch := range opt.transformationRuleBatches {
+		if i > 0 {
+			// If this is not the first batch, we need to reset
+			// the `Explored` field for the new batch.
+			g.ResetExplored()
```

Comment:
This reset operation looks quite expensive, can we use a bool flag here and combine it with `Group.Explored` / `GroupExpr.Explored` using `xor` when exploring the groups to avoid the additional traverse here? though the `Explored` field would not be that intuitive for debugging then.